File: kotlin/src/main/kotlin/DataStructures/Maps.kt
package DataStructures

/* ==================== Maps ==================== */
fun main() {
    /*
        In Kotlin, Maps use a "to" keyword to indicate a Key, Value pairing.
        There are multiple types of options to create Maps in Kotlin:

        1. mapOf()
        Create a read-only map

        Example:
            // Implicit Declaration
            val people = mapOf(1 to "Jenny", 2 to "James", 3 to "Michael", 4 to "Kenny")
            // Explicit Declaration
            val people: Map<Int, String> = mapOf<Int, String>(9 to "Zion", 2 to "Charles", 5 to "Adam")

        2. mutableMapOf()
        Creates a MutableMap. This time you can modify (add/remove) elements.

        Example:
            // Implicit Declaration
            val people = mutableMapOf(1 to "Jenny", 2 to "James", 3 to "Michael", 4 to "Kenny")
            // Explicit Declaration
            val people: MutableMap<Int, String> = mutableMapOf<Int, String>(9 to "Zion", 2 to "Charles")

        3. HashMap<K, V>()
        Creates a HashMap, which is based off the MutableMap interface. It must be defined
        explicitly.

        Example:
            // Empty HashMap
            val people = HashMap<Int, String>()
            // Initialised HashMap
            val people = HashMap<Int, String>(mapOf(1 to "Jenny", 2 to "Adam"))
    */

    // Note: Functions are the same for Maps and HashMaps
    val people1: MutableMap<Int, String> = mutableMapOf<Int, String>()
    // alternatively:
    // val people1: HashMap<Int, String> = HashMap<Int, String>()

    // 1. Add to Map: .put() or []
    people1.put(1, "Kenny")
    people1[3] = "Adam"
    people1.put(4, "Charles")
    people1[2] = "Zion"
    people1.put(5, "Xavier")
    println(people1)

    // 2. Iterate Over Map
    for ((key, value) in people1) {
        println("$key : $value")
    }

    // Alternatively...
    people1.forEach{ (key, value) ->
        println("$key : $value")
    }

    // 3. Check if Key or Value in Map: .containsKey() or .containsValue()
    println(people1.containsKey(2)) // true
    println(people1.containsKey(100)) // false
    println(people1.containsValue("Adam")) // true
    println(people1.containsValue("Tomato")) // false

    // 3. Get from Map: .get() or []
    println(people1.get(2)) // Zion
    println(people1[3]) // Adam

    // 3. A) Get all Keys: .keys (MutableSet<K>)
    val keys: MutableSet<Int> = people1.keys
    for (key in keys) {
        println(key);
    }

    // 3. B) Get all Values: .values (MutableCollection<V>)
    val values: MutableCollection<String> = people1.values
    for (value in values) {
        println(value)
    }

    // 3. C) Get Key, Value Pairs: .entries (MutableSet<MutableMap.MutableEntry<K, V>>)
    val entries: MutableSet<MutableMap.MutableEntry<Int, String>> = people1.entries
    for (entry in entries) {
        println(entry)
    }

    // 4. Edit Map: .set() or []
    people1.set(2, "ZION")
    people1[1] = "KENNY"
    println(people1)

    // 5. Remove from Map: .remove(K) or .remove(K, V)
    people1.remove(2)
    println(people1)
    people1.remove(5, "Xav")
    println(people1)
    people1.remove(5, "Xavier")
    println(people1)

}

File: cpp/DataStructures/maps.cpp
/*
 * Maps are essentially the same as HashMaps in Java/Kotlin.
 *
 * To use vectors, you need to #include <map>
*/
#include <iostream>
#include <map>

int maps() {

    std::map<int, std::string> people = {
            {2, "Adam"},
            {5, "Brian"},
            {3, "Zion"}
    };

    // 0. Iterate Over
    // Using the iterator, you can reference the key and value via:
    // iterator->first (for Key) and iterator->second (for Value)
    // The -> is basically "deferencing" the iterator to get the actual value. i->first is equivalent to writing:
    // (*i).first and i->second is equivalent to writing: (*i).second
    for (auto i = people.begin(); i != people.end(); i++) {
        std::cout << i->first << " : " << i->second << std::endl;
    }

    // 1. Insert into Map using .emplace() or .insert()
    // Four ways to insert into a map
    people.emplace(4, "Penelope"); // Using .emplace() - This is the preferred way to insert
    people[1] = "Xavier"; // Using [] operator
    people.insert({9, "Charles"}); // Using {} notation
    people.insert(std::pair<int, std::string>(6, "Danny")); // Using std::pair<type1, type2>() notation

    // 2. Get Elements
    std::cout << people.at(2) << std::endl; // .at()
    std::cout << people[3] << std::endl; // [] operator
    // Alternative method using .find() - See below section 5.

    // 3. Edit Elements using [] operator
    std::cout << people.at(9) << std::endl;
    people[9] = "CHARLES";
    std::cout << people.at(9) << std::endl;

    // 4. Remove Elements using .erase()
    people.erase(2);

    // 5. .find() Method
    /*
     * .find(k) searches the map for an element with a key equivalent to k and returns an iterator to it if found,
     * otherwise it returns an iterator to map::end.
    */
    std::map<int, std::string> mapExample = {
            {2, "Apple"},
            {3, "Banana"},
            {4, "Dragonfruit"}
    };
    // A) Change element based off iterator
    auto toChange = mapExample.find(2);
    toChange->second = "APPLE";
    std::cout << mapExample.at(2) << std::endl;
    // B) Check if element is in the map
    if (mapExample.find(10) == mapExample.end()) {
        std::cout << "Key: 10 is not in the map" << std::endl;
    }

    // 6. .contains() Method
    /*
     * .contains(k) also only checks for an element with a key equivalent to k
    */
    if (mapExample.contains(2)) {
        std::cout << "Contains key: 2 in mapExample" << std::endl;
    }

    // Sorting Maps
    /*
     * Maps store the keys in a default order. If you want to change/customise this ordering, you pass it into
     * an "Alocate" parameter when you initialise the map
    */
    std::map<int, std::string, std::greater<>> names = {
            {2, "Richard"},
            {5, "Adam"},
            {10, "Michael"},
            {4, "William"},
            {3, "Brian"}
    };

    for (auto i = names.begin(); i != names.end(); i++) {
        std::cout << i->first << " : " << i->second << std::endl;
    }

    return 0;
}